haystack/nodes/connector/crawler.py

```diff
@@ -236,16 +250,31 @@ def _write_to_files(
             else:
                 text = el.text
 
-            link_split_values = link.replace("https://", "").split("/")
-            file_name = f"{'_'.join(link_split_values)}.json"
-            file_path = output_dir / file_name
-
             data = {}
             data["meta"] = {"url": link}
             if base_url:
                 data["meta"]["base_url"] = base_url
             data["content"] = text
             document = Document.from_dict(data, id_hash_keys=id_hash_keys)
+
+            param_naming = f"{link}{text}"
+            if crawler_naming_function is not None:
+                file_name_preffix_tmp = crawler_naming_function(link, text)
+                link_split_values = (
+                    file_name_preffix_tmp.replace("https://", "")
+                    .replace("http://", "")
+                    .replace("file:/", "")
+                    .replace("file://", "")
+                    .replace("\0", "")
+                    .split("/")
+                )
+                file_name_preffix = f"{'_'.join(link_split_values)}"
+            else:
+                file_name_preffix = hashlib.md5(param_naming.encode("utf-8")).hexdigest()
```

Comment:
Don't you think that keeping as default a behavior that we know will cause error in some scenarios is a problem raiser ? For example, when we are crawling the web we don't know the links we are going to crawl. So the lenght of the link is not a pre crawling guess easy to make by users, as a result many users could have issues with this, raising new issues on repo and even on Deepset Cloud support. I used this point of view to change default behavior. Reduce pressure on support team and keep user experience better (less possible errors by default).

What you think ? Should I change ?

Reply:
The unaccepted chars into a file name are:

*Windows  < > : " / \ | ? *
*UNIX  \0 /

What do you think about a regex to replace then on the file name ? So haystack would never throw an error. Assuming haystack don't want to prioritize UNIX systems.

Reply:
If we use something like the following code we would avoid unnecessary errors on Windows, Linux, and Mac.

```python
re.sub("[<>:'/\\|?*\0 ]","_",file_name_preffix_tmp)
```

Reply:
These are all valid points! I'll take them one by one.

> Don't you think that keeping as default a behavior that we know will cause error in some scenarios is a problem raiser ? For example, when we are crawling the web we don't know the links we are going to crawl.

I don't think it's always true that users don't know how the links are going to look like. Of course if you're crawling the entire Internet, then it's true; but often this Crawler is used to crawl knowledge bases and wikis, to perform semantic search within them. And such website tend to have very informative and simple URLs, which would be a shame to waste.

I believe we should optimize for this usecase: this crawler is very simple and we would like it to stay simple and intuitive. Internet-wide crawling is better done with more powerful tools anyway :slightly_smiling_face: 

Consider also that if we alter the behavior of the Crawler radically, this would count as a breaking change and users upgrading from older versions of Haystack might meet with unexpected problems.

> What do you think about a regex to replace then on the file name ? So haystack would never throw an error. 

In some cases, changing the filename arbitrarily to avoid an error can cause more trouble. 

Imagine if a Windows user was crawling two pages called `www.mywebsite.com/usage_of_<=` and `www.mywebsite.com/usage_of_<`. That's a small example, but your regex would make these two links identical and raise no warning. The user will simply miss a page. What if some important data was in it? How are they going to notice that the page is missing? If you just threw an exception, at least the user would be aware.

My point is that we should give them a very simple default behavior, suitable only for the most common and basic cases, and explain clearly the errors when they arise, so they can take only the steps necessary to correct them for their specific website, OS, etc. Users of Haystack are developers as well, we don't need to hand-hold them. They know their usecase and their limitations better than us :slightly_smiling_face: 

----

Of course, please reply to all of this if you're still unconvinced! :blush:

Reply:
>Imagine if a Windows user was crawling two pages called `www.mywebsite.com/usage_of_<= `and `www.mywebsite.com/usage_of_<`, they are one char different in size. There are possibilities to handle these errors, on naming, but It would insert more complexity on the code, like dict loops and some checks. But I got the point, this is one of the reasons I proposed the MD5 hash on link + text, similar to audio files, where the default is MD5 of text.

The two pages will be saved under different file names
> Consider also that if we alter the behavior of the Crawler radically, this would count as a breaking change and users upgrading from older versions of Haystack might meet with unexpected problems.
I don't see this as a breaking change because the file name is not processed by Haystack, nowhere else. I assured this reading the source code. If any user is making external usage of crawled files, referencing them by name, this would be a non domain (NLP) usage of Haystack (like using it as a common crawler).

> Imagine if a Windows user was crawling two pages called www.mywebsite.com/usage_of_<= and www.mywebsite.com/usage_of_<. That's a small example, but your regex would make these two links identical and raise no warning. The user will simply miss a page. What if some important data was in it? How are they going to notice that the page is missing? If you just threw an exception, at least the user would be aware.
That's the main reason I proposed for a MD5 of link + text. This will give us almost always (there exist MD5 collisions, so I will not say always) as the default behavior. So the naming function would be the total responsibility of the user. And framework won't be responsible for any errors. 

> Users of Haystack are developers as well, we don't need to hand-hold them. They know their usecase and their limitations better than us slightly_smiling_face
Yeah, you are right. I usually develop for end users which are non-tech ones. The hand-hold is not needed. But as on my first intense usage of Haystack I noticed the Crawler hanging over unexpectedly when crawling a law news site (which we use to enrich our NLP models) - www.migalhas.com.br I tried to minimize this kind of error.